test/test_images/probe_helper/probe/handler.go

```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package probe
+
+import (
+	"context"
+)
+
+/*
+
+The following steps should be taken when implementing a new type of probe.
+
+1. Define a forward probe handler struct which implements the Handler interface,
+  and define a constructor for that probe Handler object.
+2. Define a receiver probe handler struct which implements the Handler interface,
+	and define a constructor for that probe Handler object.
+3. Ensure that the ChannelID for paired forward/receive probe Handlers are equal.
+4. Add the mapping between forward probe type and forward Handler constructor to
+	the forwardProbeConstructors map in probe/helper.go
+5. Add the mapping between receiver probe type and receiver Handler constructor to
+  the receiveProbeConstructors map in probe/helper.go
+
+*/
+
+// Handler is the interface which probe objects should implement.
+type Handler interface {
+	// ChannelID represents the key of the receiver channel which the probe helper
+	// should wait on after forwarding a probe request. You should ensure that
+	// each distinct probe request maps to a unique ChannelID, and that when a
+	// forward probe is matched with a receiver probe, that both handlers produce
+	// the same ChannelID.
+	ChannelID() string
```

Comment:
Thinking about this a bit more, I think we can hide the details of the channel. To be more specific, I am considering something like this:
```
type Handler interface {
     Forward(ctx context.Context, event Event)
     Receive(ctx context.Context, event Event)
}
```

The channel becomes an implementation detail  of each handler (some handlers don't even need the channel maybe). There will a single handler for each event type.

Reply:
I agree with creating a static handler for each event type, that's definitely a better idea than what I currently have.

I'm not sure it is a good idea to have the channel be an implementation detail of each handler. This would mean duplicating the channel creation, waiting, and deletion for each handler. This could increase the likelihood of subtle channel manipulation mistakes between handlers, but also would dirty up each of the handler implementations. I think one of the main points of this refactoring is precisely about gathering the common logic between probes (like the channel manipulation) in a generic handler.

Reply:
I modified the Handler to be as you suggested, let me know what you think.